src/ark/utils/deepcell_service_utils.py

```diff
@@ -99,85 +176,41 @@ def create_deepcell_output(deepcell_input_dir, deepcell_output_dir, fovs=None,
 
     print(f'Processing tiffs in {len(fov_groups)} batches...')
 
-    # yes this is function, don't worry about it
-    # long story short, too many args to pass if function not in local scope
-    # i.e easier to map fov_groups
-    def _zip_run_extract(fov_group, group_index):
-        # define the location of the zip file for our fovs
-        zip_path = os.path.join(deepcell_input_dir, f'fovs_batch_{group_index + 1}.zip')
-
-        if os.path.isfile(zip_path):
-            warnings.warn(f'{zip_path} will be overwritten')
-
-        # write all files to the zip file
-        print('Zipping preprocessed tiff files.')
-
-        def zip_write(zip_path):
-            with ZipFile(zip_path, 'w', compression=ZIP_DEFLATED) as zipObj:
-                for fov in fov_group:
-                    # file has .tiff extension
-                    basename = fov + '.tiff'
-                    filename = os.path.join(deepcell_input_dir, basename)
-                    zipObj.write(filename, basename)
-
-        zip_write(zip_path)
-
-        # pass the zip file to deepcell.org
-        print('Uploading files to DeepCell server.')
-        status = run_deepcell_direct(
-            zip_path, deepcell_output_dir, host, job_type, scale, timeout
-        )
-
-        # ensure execution is halted if run_deepcell_direct returned non-zero exit code
-        if status != 0:
-            print("The following FOVs could not be processed: %s" % ','.join(fov_group))
-            return
+    for batch_num, fov_group in enumerate(fov_groups):
+        # create zipped input files
+        input_zip_path = zip_input_files(deepcell_input_dir, fov_group, batch_num)
 
-        # extract the .tif output
-        print("Extracting tif files from DeepCell response.")
-        zip_names = io_utils.list_files(deepcell_output_dir, substrs=[".zip"])
+        # add timeout limit
+        batch_filename = Path(input_zip_path).name
+        output_zip_path = os.path.join(deepcell_output_dir, f"deepcell_response_" + batch_filename)
+        while not os.path.exists(output_zip_path):
+            # upload to deepcell
```

Comment:
I need to add some kind of timeout here to avoid an endless loop if the deepcell upload continuously fails. I haven't run segmentation on enough data to know a reasonable time limit. Any suggestions? (Keep in mind a single successful iteration will consist of uploading the files to deepcell, running the prediction script, and extracting the tiffs from the deepcell output zip file.) @alex-l-kong @srivarra

Reply:
Because cohorts may be very large, we don't want a single batch of 5 FOVs to be holding up the line. Some leeway is fine just in case the network is slow, but since we'll be including the FOV names in these batches, it should not be difficult for the user to figure out which FOVs failed, if any, and which ones to re-run.

From my experience a FOV batch should not take exceptionally long to process. I'd say no more than 2-3 mins, and depending how long 2048x2048 images take, I might even drop down to 1-1.5 mins.